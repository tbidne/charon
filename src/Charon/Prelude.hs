{-# LANGUAGE CPP #-}
{-# LANGUAGE ImplicitParams #-}
{-# LANGUAGE MagicHash #-}
{-# LANGUAGE QuasiQuotes #-}

-- | Custom prelude.
module Charon.Prelude
  ( module X,

    -- * Text
    bsToStr,
    bsToStrLenient,
    showt,
    displayExceptiont,
    decodeDisplayExT,

    -- * Display Utils
    (<+>),
    vsep,
    punctuate,
    line,

    -- * Debug
    todo,

    -- * Optics
    packed,
    unpacked,

    -- ** Path literals
    pathDotTrash,
    pathCharon,

    -- * Misc
    charonPath,
    doesAnyPathExist,
    doesAnyPathNotExist,
    throwLeft,
    usingReaderT,
  )
where

import Control.Applicative as X
  ( Alternative ((<|>)),
    Applicative (liftA2, pure, (<*>)),
    (*>),
  )
import Control.Category as X (Category ((.)), (>>>))
import Control.DeepSeq as X (NFData)
import Control.Exception as X (IOException)
import Control.Exception.Utils as X
  ( StringException (MkStringException),
    catchSync,
    catchesSync,
    exitFailure,
    throwString,
    throwText,
    trySync,
  )
import Control.Monad as X
  ( Monad ((>>=)),
    join,
    unless,
    void,
    when,
    (<=<),
    (=<<),
    (>=>),
  )
import Control.Monad.Catch as X
  ( Exception (displayException),
    Handler (Handler),
    MonadCatch,
    MonadMask,
    MonadThrow,
    SomeException,
    bracket,
    catch,
    catches,
    finally,
    throwM,
    try,
  )
import Control.Monad.Fail as X (MonadFail (fail))
import Control.Monad.IO.Class as X (MonadIO (liftIO))
import Control.Monad.Reader as X
  ( MonadReader (ask),
    ReaderT,
    asks,
    local,
    runReaderT,
  )
import Data.Bifunctor as X (Bifunctor (bimap, first, second))
import Data.Bool as X (Bool (False, True), not, otherwise, (&&), (||))
import Data.ByteString as X (ByteString)
import Data.Bytes as X (Bytes (MkBytes), Size (B), _MkBytes)
import Data.Char as X (Char)
import Data.Coerce as X (coerce)
import Data.Either as X (Either (Left, Right), either)
import Data.Eq as X (Eq ((/=), (==)))
import Data.Foldable as X
  ( Foldable (fold, foldMap', foldl', foldr, foldr1, length),
    foldlM,
    for_,
    null,
    sequenceA_,
    traverse_,
  )
import Data.Foldable1 as X
  ( Foldable1 (foldMap1),
    foldlM1,
    foldlMapM1,
  )
import Data.Function as X (const, flip, id, ($))
import Data.Functor as X (Functor (fmap), ($>), (<$>), (<&>))
import Data.HashMap.Strict as X (HashMap)
import Data.HashSet as X (HashSet)
import Data.Hashable as X (Hashable (hashWithSalt))
import Data.Int as X (Int)
import Data.Kind as X (Constraint, Type)
import Data.List as X (filter, zipWith, (++))
import Data.List.NonEmpty as X (NonEmpty ((:|)))
import Data.Maybe as X (Maybe (Just, Nothing), fromMaybe, maybe)
import Data.Monoid as X (Monoid (mconcat, mempty))
import Data.Ord as X
  ( Ord (compare, (<), (<=), (>), (>=)),
    Ordering (EQ, GT, LT),
    min,
  )
import Data.Proxy as X (Proxy (Proxy))
import Data.Semigroup as X (Semigroup ((<>)))
import Data.Sequence as X (Seq ((:<|), (:|>)))
import Data.Sequence.NonEmpty as X (NESeq ((:<||), (:||>)))
import Data.String as X (IsString (fromString), String)
import Data.Text as X (Text)
import Data.Text qualified as T
import Data.Text.Builder.Linear (Builder)
import Data.Text.Display as X (Display (displayBuilder), display)
import Data.Traversable as X (for, traverse)
import Data.Tuple as X (curry, fst, uncurry)
#if MIN_VERSION_base(4, 17, 0)
import Data.Type.Equality as X (type (~))
#endif
import Data.Vector as X (Vector)
import Data.Word as X (Word16, Word8)
import Effects.Concurrent.Async as X (MonadAsync)
import Effects.Concurrent.Thread as X (MonadThread)
import Effects.FileSystem.FileReader as X
  ( MonadFileReader (readBinaryFile),
    decodeUtf8,
    decodeUtf8Lenient,
    readFileUtf8ThrowM,
  )
import Effects.FileSystem.FileWriter as X
  ( MonadFileWriter (appendBinaryFile, writeBinaryFile),
    encodeUtf8,
  )
import Effects.FileSystem.HandleWriter as X
  ( MonadHandleWriter
      ( hClose,
        hFlush,
        hPut,
        openBinaryFile
      ),
  )
import Effects.FileSystem.PathReader as X
  ( MonadPathReader
      ( canonicalizePath,
        doesDirectoryExist,
        doesFileExist,
        doesPathExist,
        getFileSize,
        getHomeDirectory,
        getTemporaryDirectory,
        listDirectory,
        makeAbsolute
      ),
    PathType
      ( PathTypeDirectory,
        PathTypeFile,
        PathTypeOther,
        PathTypeSymbolicLink
      ),
    doesSymbolicLinkExist,
    getXdgConfig,
  )
import Effects.FileSystem.PathWriter as X
  ( MonadPathWriter
      ( createDirectoryIfMissing,
        removeDirectoryRecursive,
        removePathForcibly,
        renameDirectory,
        renameFile
      ),
  )
import Effects.Haskeline as X (MonadHaskeline, getInputChar, getInputLine)
import Effects.IORef as X
  ( IORef,
    MonadIORef,
    modifyIORef',
    newIORef,
    readIORef,
    writeIORef,
  )
import Effects.Logger as X
  ( LogLevel (LevelDebug, LevelError, LevelInfo, LevelWarn),
    MonadLogger (monadLoggerLog),
    levelFatal,
    logDebug,
    logError,
    logFatal,
    logInfo,
    logWarn,
  )
import Effects.Logger.Namespace as X
  ( HasNamespace,
    MonadLoggerNS,
    Namespace,
    addNamespace,
  )
import Effects.Optparse as X (MonadOptparse (execParser))
#if !WINDOWS
import Effects.System.Posix.Files as X (MonadPosixFiles)
#endif
import Effects.System.PosixCompat.Files as X (MonadPosixCompatFiles)
import Effects.System.Terminal as X
  ( MonadTerminal (putStr, putStrLn),
    print,
    putTextLn,
  )
import Effects.Time as X (MonadTime)
import FileSystem.OsPath as X
  ( OsPath,
    decodeDisplayEx,
    decodeLenient,
    encodeValid,
    encodeValidThrowM,
    osp,
    ospPathSep,
    (<.>),
    (</>),
  )
import FileSystem.OsString as X
  ( OsString,
    osstr,
  )
import GHC.Enum as X (Bounded (maxBound, minBound), Enum (toEnum))
import GHC.Err as X (error, undefined)
import GHC.Exception (errorCallWithCallStackException)
import GHC.Exts (RuntimeRep, TYPE, raise#)
import GHC.Float as X (Double)
import GHC.Generics as X (Generic)
import GHC.Integer as X (Integer)
import GHC.Natural as X (Natural)
import GHC.Num as X (Num ((*), (+), (-)))
import GHC.Real as X (Integral (div), even, fromIntegral, realToFrac)
import GHC.Stack as X
  ( CallStack,
    HasCallStack,
    callStack,
    prettyCallStack,
  )
import Numeric.Convert.Integer as X (FromInteger, fromℤ, toℤ)
import Numeric.Convert.Rational as X (FromRational, fromℚ)
import Optics.Core as X
  ( A_Getter,
    A_Lens,
    A_Setter,
    AffineTraversal',
    Getter,
    Is,
    Iso',
    LabelOptic (labelOptic),
    LabelOptic',
    Lens,
    Lens',
    Optic',
    Prism',
    iso,
    lens,
    lensVL,
    over',
    preview,
    prism,
    review,
    set',
    to,
    view,
    (%),
    (%?),
    (&),
    (.~),
    (^.),
    (^?),
    _1,
    _2,
    _3,
    _4,
    _Just,
  )
import Optics.Core.Extras as X (is)
import Optics.TH as X
  ( generateUpdateableOptics,
    makeFieldLabelsNoPrefix,
    makeFieldLabelsWith,
    makePrisms,
    noPrefixFieldLabels,
  )
import PathSize as X (findLargestPaths)
import PathSize.Utils as X (MonadPosixFilesC)
import System.Exit as X (ExitCode (ExitFailure, ExitSuccess))
import System.IO as X
  ( FilePath,
    Handle,
    IO,
    IOMode (AppendMode),
  )
import Text.Show as X (Show (show))

showt :: (Show a) => a -> Text
showt = T.pack . show

displayExceptiont :: (Exception e) => e -> Text
displayExceptiont = T.pack . displayException

bsToStr :: ByteString -> String
bsToStr = either displayException T.unpack . decodeUtf8

bsToStrLenient :: ByteString -> String
bsToStrLenient = T.unpack . decodeUtf8Lenient

-- Vendoring optics-extra Data.Text.Strict.Optics

packed :: Iso' String Text
packed = iso T.pack T.unpack
{-# INLINE packed #-}

unpacked :: Iso' Text String
unpacked = iso T.unpack T.pack
{-# INLINE unpacked #-}

usingReaderT :: b -> ReaderT b m a -> m a
usingReaderT = flip runReaderT

pathCharon :: OsPath
pathCharon = [osp|charon|]

pathDotTrash :: OsPath
pathDotTrash = [osp|.trash|]

doesAnyPathExist ::
  ( HasCallStack,
    MonadCatch m,
    MonadPathReader m
  ) =>
  OsPath ->
  m Bool
doesAnyPathExist p = do
  symlinkExists <- doesSymbolicLinkExist p
  if symlinkExists
    then pure True
    else doesPathExist p

doesAnyPathNotExist ::
  ( HasCallStack,
    MonadCatch m,
    MonadPathReader m
  ) =>
  OsPath ->
  m Bool
doesAnyPathNotExist = fmap not . doesAnyPathExist

decodeDisplayExT :: OsPath -> Text
decodeDisplayExT = T.pack . decodeDisplayEx

vsep :: [Builder] -> Builder
vsep = concatWith (\x y -> x <> line <> y)

punctuate :: Builder -> [Builder] -> [Builder]
punctuate _ [] = []
punctuate _ [x] = [x]
punctuate p (x : xs) = x <> p : punctuate p xs

(<+>) :: Builder -> Builder -> Builder
x <+> y = x <> " " <> y

line :: Builder
line = "\n"

-- vendored from prettyprinter, for text's Builder
concatWith :: (Foldable t) => (Builder -> Builder -> Builder) -> t Builder -> Builder
concatWith f ds
  | null ds = mempty
  | otherwise = foldr1 f ds

-- | Placeholder for unwritten code.
todo :: forall {r :: RuntimeRep} (a :: TYPE r). (HasCallStack) => a
todo = raise# (errorCallWithCallStackException "Prelude.todo: not yet implemented" ?callStack)
{-# WARNING todo "todo remains in code" #-}

charonPath :: OsPath
charonPath = [osp|charon|]

throwLeft :: forall m e a. (Exception e, HasCallStack, MonadThrow m) => Either e a -> m a
throwLeft (Right x) = pure x
throwLeft (Left e) = throwM e
