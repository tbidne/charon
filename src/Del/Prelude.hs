-- | Custom prelude.
--
-- @since 0.1
module Del.Prelude
  ( module X,
  )
where

import Control.Applicative as X
  ( Alternative ((<|>)),
    Applicative (pure, (<*>)),
    (*>),
  )
import Control.DeepSeq as X (NFData)
import Control.Exception as X
  ( Exception (displayException, fromException, toException),
    SomeAsyncException (SomeAsyncException),
    SomeException,
    bracket,
    catch,
    finally,
    throwIO,
  )
import Control.Monad as X (Monad ((>>=)), join, (<=<), (>=>))
import Control.Monad.Fail as X (MonadFail (fail))
import Data.Bool as X (Bool (False, True), not, otherwise, (&&))
import Data.ByteString as X (ByteString)
import Data.Either as X (Either (Left, Right))
import Data.Eq as X (Eq ((/=), (==)))
import Data.Foldable as X (Foldable (foldl', foldr, length), for_)
import Data.Function as X (const, flip, ($), (.))
import Data.Functor as X (Functor (fmap), ($>), (<$>))
import Data.HashMap.Strict as X (HashMap)
import Data.HashSet as X (HashSet)
import Data.Hashable as X (Hashable (hashWithSalt))
import Data.IORef as X (IORef, modifyIORef', newIORef, readIORef)
import Data.Int as X (Int)
import Data.Kind as X (Type)
import Data.List as X (zipWith)
import Data.List.NonEmpty as X (NonEmpty ((:|)))
import Data.Maybe as X (Maybe (Just, Nothing), maybe)
import Data.Monoid as X (Monoid (mconcat, mempty))
import Data.Ord as X (Ord (compare, (<=), (>), (>=)), Ordering (EQ, GT, LT))
import Data.Proxy as X (Proxy (Proxy))
import Data.Semigroup as X (Semigroup ((<>)))
import Data.String as X (IsString (fromString), String)
import Data.Text as X (Text)
import Data.Traversable as X (traverse)
import Data.Tuple as X (curry, uncurry)
import Data.Vector as X (Vector)
import Data.Word as X (Word16)
import GHC.Enum as X (Bounded (maxBound, minBound))
import GHC.Err as X (error, undefined)
import GHC.Float as X (Double)
import GHC.Generics as X (Generic)
import GHC.Integer as X (Integer)
import GHC.Natural as X (Natural)
import GHC.Num as X (Num ((+), (-)))
import GHC.Real as X (fromIntegral)
import GHC.Stack as X (HasCallStack)
import Optics.Core as X (Iso', iso, review, view, (%), (^.), _1, _2)
import Prettyprinter as X
  ( Pretty (pretty),
    layoutCompact,
    line,
    vsep,
    (<+>),
  )
import Prettyprinter.Render.Text as X (renderStrict)
import System.Directory as X (createDirectoryIfMissing)
import System.Exit as X (exitFailure)
import System.FilePath as X ((</>))
import System.IO as X
  ( BufferMode (NoBuffering),
    FilePath,
    IO,
    print,
    putStr,
    putStrLn,
  )
import Text.Show as X (Show (show))
