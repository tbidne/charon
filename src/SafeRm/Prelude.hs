-- | Custom prelude.
--
-- @since 0.1
module SafeRm.Prelude
  ( module X,

    -- * Text
    showt,
  )
where

import Control.Applicative as X
  ( Alternative ((<|>)),
    Applicative (pure, (<*>)),
    (*>),
  )
import Control.DeepSeq as X (NFData)
import Control.Monad as X
  ( Monad ((>>=)),
    join,
    unless,
    void,
    when,
    (<=<),
    (=<<),
    (>=>),
  )
import Control.Monad.Fail as X (MonadFail (fail))
import Control.Monad.IO.Class as X (MonadIO (liftIO))
import Control.Monad.Logger as X
  ( LogLevel (LevelDebug, LevelError, LevelInfo, LevelWarn),
    MonadLogger (monadLoggerLog),
    logDebug,
    logError,
    logInfo,
    logWarn,
  )
import Control.Monad.Reader as X
  ( MonadReader (ask),
    ReaderT,
    asks,
    local,
    runReaderT,
  )
import Data.Bifunctor as X (Bifunctor (bimap))
import Data.Bool as X (Bool (False, True), not, otherwise, (&&), (||))
import Data.ByteString as X (ByteString)
import Data.Char as X (Char)
import Data.Either as X (Either (Left, Right), either)
import Data.Eq as X (Eq ((/=), (==)))
import Data.Foldable as X
  ( Foldable (foldMap', foldl', foldr, length),
    for_,
    null,
    sequenceA_,
  )
import Data.Function as X (const, flip, id, ($), (.))
import Data.Functor as X (Functor (fmap), ($>), (<$>), (<&>))
import Data.HashMap.Strict as X (HashMap)
import Data.HashSet as X (HashSet)
import Data.Hashable as X (Hashable (hashWithSalt))
import Data.Int as X (Int)
import Data.Kind as X (Constraint, Type)
import Data.List as X (filter, zipWith)
import Data.List.NonEmpty as X (NonEmpty ((:|)))
import Data.Maybe as X (Maybe (Just, Nothing), fromMaybe, maybe)
import Data.Monoid as X (Monoid (mconcat, mempty))
import Data.Ord as X
  ( Ord (compare, (<), (<=), (>), (>=)),
    Ordering (EQ, GT, LT),
    min,
  )
import Data.Proxy as X (Proxy (Proxy))
import Data.Semigroup as X (Semigroup ((<>)))
import Data.Sequence as X (Seq, (<|), (|>))
import Data.String as X (IsString (fromString), String)
import Data.Text as X (Text)
import Data.Text qualified as T
import Data.Traversable as X (traverse)
import Data.Tuple as X (curry, uncurry)
import Data.Vector as X (Vector)
import Data.Word as X (Word16)
import GHC.Enum as X (Bounded (maxBound, minBound), Enum (toEnum))
import GHC.Err as X (error, undefined)
import GHC.Float as X (Double)
import GHC.Generics as X (Generic)
import GHC.Integer as X (Integer)
import GHC.Natural as X (Natural)
import GHC.Num as X (Num ((+), (-)))
import GHC.Real as X (even, fromIntegral)
import GHC.Stack as X
  ( CallStack,
    HasCallStack,
    callStack,
    prettyCallStack,
  )
import Optics.Core as X
  ( A_Getter,
    A_Lens,
    A_Setter,
    AffineTraversal',
    Is,
    Iso',
    LabelOptic (labelOptic),
    LabelOptic',
    Lens,
    Lens',
    Optic',
    Prism',
    iso,
    lens,
    over',
    preview,
    prism,
    review,
    set',
    view,
    (%),
    (%?),
    (^.),
    (^?),
    _1,
    _2,
    _3,
    _4,
    _Just,
  )
import Optics.TH as X (makeFieldLabelsNoPrefix, makePrisms)
import Prettyprinter as X
  ( Pretty (pretty),
    layoutCompact,
    line,
    vsep,
    (<+>),
  )
import Prettyprinter.Render.Text as X (renderStrict)
import System.Exit as X (exitFailure)
import System.FilePath as X ((</>))
import System.IO as X
  ( BufferMode (NoBuffering),
    FilePath,
    Handle,
    IO,
    IOMode (AppendMode),
  )
import Text.Show as X (Show (show))
import UnliftIO as X (MonadUnliftIO)
import UnliftIO.Exception as X
  ( Exception (displayException, fromException, toException),
    SomeAsyncException (SomeAsyncException),
    SomeException,
    bracket,
    bracket_,
    catch,
    catchAny,
    finally,
    handleAny,
    throwIO,
    throwString,
    try,
    tryAny,
  )
import UnliftIO.IORef as X (IORef, modifyIORef', newIORef, readIORef)

-- | @since 0.1
showt :: Show a => a -> Text
showt = T.pack . show
