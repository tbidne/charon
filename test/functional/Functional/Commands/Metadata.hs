{-# LANGUAGE CPP #-}
{-# LANGUAGE QuasiQuotes #-}

-- | Tests for m command.
module Functional.Commands.Metadata
  ( tests,
  )
where

import Functional.Prelude

tests :: IO TestEnv -> TestTree
tests testEnv =
  testGroup
    "Metadata Command"
    [ metadata testEnv',
      empty testEnv'
    ]
  where
    testEnv' = appendTestDir "metadata" <$> testEnv

metadata :: IO TestEnv -> TestTree
metadata getTestEnv = testCase "Prints metadata" $ do
  testEnv <- getTestEnv
  usingReaderT testEnv $ appendTestDirM "metadata" $ do
    testDir <- getTestDir

    let filesToDelete = (testDir </>!) <$> ["f1", "f2", "f3"]
        dirsToDelete = (testDir </>!) <$> ["dir1", "dir2", "dir4"]
        fileLinkToDelete = testDir </> [osp|file-link|]
        dirLinkToDelete = testDir </> [osp|dir-link|]
        linksToDelete = [fileLinkToDelete, dirLinkToDelete]

    delArgList <- withSrArgsPathsM ["delete"] (filesToDelete <> dirsToDelete <> linksToDelete)

    -- setup
    clearDirectory testDir
    -- test w/ a nested dir
    createDirectories ((testDir </>!) <$> ["dir1", "dir2", "dir2/dir3", "dir4"])
    -- test w/ a file in dir
    createFiles ((testDir </>! "dir2/dir3/foo") : filesToDelete)
    createSymlinks [F fileLinkToDelete, D dirLinkToDelete, F $ testDir </>! "dir4" </>! "link"]
    assertPathsExist (filesToDelete ++ dirsToDelete)

    runCharon delArgList

    -- file assertions
    assertPathsDoNotExist (filesToDelete ++ dirsToDelete)
    assertSymlinksDoNotExist linksToDelete

    -- trash structure assertions
    delExpectedIdxSet <-
      mkPathDataSetM
        [ ("f1", PathTypeFile, 5),
          ("f2", PathTypeFile, 5),
          ("f3", PathTypeFile, 5),
          ("dir1", PathTypeDirectory, 5),
          ("dir2", PathTypeDirectory, 15),
          ("dir4", PathTypeDirectory, 10),
          ("dir-link", PathTypeSymbolicLink, 5),
          ("file-link", PathTypeSymbolicLink, 5)
        ]

    (delIdxSet, delMetadata) <- runIndexMetadataM
    assertSetEq delExpectedIdxSet delIdxSet
    delExpectedMetadata @=? delMetadata
    assertFdoDirectorySizesM ["dir1", "dir2", "dir4"]

    -- METADATA

    metaArgList <- withSrArgsM ["metadata"]
    (metadataResult, _) <- captureCharonLogs metaArgList

    -- assert nothing changed
    assertPathsDoNotExist (filesToDelete ++ dirsToDelete)

    assertMatches expectedMetadata metadataResult

    -- trash structure assertions
    (metadataIdxSet, metadatMetadata) <- runIndexMetadataM
    assertSetEq delExpectedIdxSet metadataIdxSet
    delExpectedMetadata @=? metadatMetadata
    assertFdoDirectorySizesM ["dir1", "dir2", "dir4"]
  where
    delExpectedMetadata = mkMetadata 8 7 0 55

{- ORMOLU_DISABLE -}

    expectedMetadata =
      Exact
        <$> [ "Entries:      8",
              "Total Files:  7",
              "Log size:     0.00B",
#if WINDOWS
              "Size:         0.00B",
#else
              "Size:         55.00B",
#endif
              ""
            ]
{- ORMOLU_ENABLE -}

empty :: IO TestEnv -> TestTree
empty getTestEnv = testCase "Prints empty metadata" $ do
  testEnv <- getTestEnv
  usingReaderT testEnv $ appendTestDirM "emptySucceeds" $ do
    testDir <- getTestDir

    let trashDir = testDir </>! ".trash"

    createDirectories [testDir, trashDir, trashDir </>! "info", trashDir </>! "files"]
    createFiles [trashDir </>! "log"]

    metaArgList <- withSrArgsM ["metadata"]
    (result, _) <- captureCharonLogs metaArgList

    assertMatches expectedTerminal result
  where
    expectedTerminal =
      Exact
        <$> [ "Entries:      0",
              "Total Files:  0",
              "Log size:     0.00B",
              "Size:         0.00B",
              ""
            ]
