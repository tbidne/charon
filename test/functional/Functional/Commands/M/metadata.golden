TERMINAL LIST
Type:      Directory
Name:      dir1
Original:  <dir>/m1/dir1
Created:   2020-05-31 12:00:00 

Type:      Directory
Name:      dir2
Original:  <dir>/m1/dir2
Created:   2020-05-31 12:00:00 

Type:      File
Name:      f1
Original:  <dir>/m1/f1
Created:   2020-05-31 12:00:00 

Type:      File
Name:      f2
Original:  <dir>/m1/f2
Created:   2020-05-31 12:00:00 

Type:      File
Name:      f3
Original:  <dir>/m1/f3
Created:   2020-05-31 12:00:00 
Entries:      5
Total Files:  4
Log size:     0.00B
Size:         25.00B 


TERMINAL METADATA
Entries:      5
Total Files:  4
Log size:     0.00B
Size:         25.00B 


LOGS METADATA
[2020-05-31 12:00:00][functional.getMetadata][Debug][src/SafeRm.hs] Trash home: <dir>/m1/.trash
[2020-05-31 12:00:00][functional.getMetadata.toMetadata.readIndex][Debug][src/SafeRm/Data/Index.hs] Index path: <dir>/m1/.trash/.index.csv
[2020-05-31 12:00:00][functional.getMetadata.toMetadata.readIndex.readIndexWithFold][Debug][src/SafeRm/Data/Index.hs] Found: MkPathData {pathType = PathTypeFile, fileName = MkPathI {unPathI = "f1"}, originalPath = MkPathI {unPathI = "<dir>/m1/f1"}, created = MkTimestamp {unTimestamp = 2020-05-31 12:00:00}}
[2020-05-31 12:00:00][functional.getMetadata.toMetadata.readIndex.readIndexWithFold][Debug][src/SafeRm/Data/Index.hs] Found: MkPathData {pathType = PathTypeFile, fileName = MkPathI {unPathI = "f3"}, originalPath = MkPathI {unPathI = "<dir>/m1/f3"}, created = MkTimestamp {unTimestamp = 2020-05-31 12:00:00}}
[2020-05-31 12:00:00][functional.getMetadata.toMetadata.readIndex.readIndexWithFold][Debug][src/SafeRm/Data/Index.hs] Found: MkPathData {pathType = PathTypeDirectory, fileName = MkPathI {unPathI = "dir1"}, originalPath = MkPathI {unPathI = "<dir>/m1/dir1"}, created = MkTimestamp {unTimestamp = 2020-05-31 12:00:00}}
[2020-05-31 12:00:00][functional.getMetadata.toMetadata.readIndex.readIndexWithFold][Debug][src/SafeRm/Data/Index.hs] Found: MkPathData {pathType = PathTypeFile, fileName = MkPathI {unPathI = "f2"}, originalPath = MkPathI {unPathI = "<dir>/m1/f2"}, created = MkTimestamp {unTimestamp = 2020-05-31 12:00:00}}
[2020-05-31 12:00:00][functional.getMetadata.toMetadata.readIndex.readIndexWithFold][Debug][src/SafeRm/Data/Index.hs] Found: MkPathData {pathType = PathTypeDirectory, fileName = MkPathI {unPathI = "dir2"}, originalPath = MkPathI {unPathI = "<dir>/m1/dir2"}, created = MkTimestamp {unTimestamp = 2020-05-31 12:00:00}}
[2020-05-31 12:00:00][functional.getMetadata.toMetadata][Debug][src/SafeRm/Data/Metadata.hs] Index size: 5
[2020-05-31 12:00:00][functional.getMetadata.toMetadata][Debug][src/SafeRm/Data/Metadata.hs] Num entries: 5
[2020-05-31 12:00:00][functional.getMetadata.toMetadata][Debug][src/SafeRm/Data/Metadata.hs] Log does not exist
[2020-05-31 12:00:00][functional.getMetadata.toMetadata][Debug][src/SafeRm/Data/Metadata.hs] Num all files: 4
[2020-05-31 12:00:00][functional.getMetadata.toMetadata][Debug][src/SafeRm/Data/Metadata.hs] Total size: MkSomeSize SB (MkBytes 25.0)
